### DISCUSSION TITLE
error-stack: Overhaul of the "related-errors" feature

### DISCUSSION THREAD

**Discussion by TimDiekmann** [hashintel/hash#2377] [error-stack (Ideas)]
`error-stack` has the concept of "related-errors". This means it's possible to have multiple frames living next to each other. The current approach, however, has some downsides:
- It's not possible to convert a `Report` into `Error` as there may be multiple sources of an error, which the `Error` trait does not support. This is especially annoying if libraries want to use `error-stack` but downstream crates don't (for whatever reason) as `Report` cannot implement `Error`.
- `Report::current_context` only returns the first context, not a list of contexts, if multiple contexts at the same level exist. (The current implementation contains a bug in addition insofar that if adding a context and an attachment with the same type the attachment may be returned instead)
- `Report::frames` and similar functions return a linearized iterator of all frames in the order in which they would be printed, however, there is no real distinction between  a related error and a non-related error
- Generally, every function needs to take a potentially related context into account (thus we e.g. removed `source` in favor of `sources`). In most use cases, however, there are no related errors.
- Constructing "related-errors" is hard as we want to ensure, that at least one context is available, thus, `FromIterator` cannot be implemented on `Report`.

"related-errors" are a great feature and have great potential when processing inside of a loop happens, but in all other cases, it's more of an annoyance.

I was thinking a lot about how to fix the above errors. Generally, I came up with two potential solutions. Both ideas revert most of the API changes done in #747 but all above points should be resolved. However, as we really want to continue supporting "related-errors", I propose to

1. either build-in functionality to support `Report<Vec<C>>`. Methods, currently returning potentially multiple things (like `sources`) would only return `T` in `Report<T>` (either `C: Context` or `Vec<C: Context>`). Sadly, this is not as simple as it sounds. Currently, we rely on `C: Context` in `Report<C>`, with `Report<Vec<C>>` this is not the case anymore. We eventually want to remove the `Context` trait as soon as `core::error::Error` has been stabilized, so it's not possible to implement `Context` for `Vec<C: Context>`. As an alternative, a `MultiContext` struct could be provided with built-in support from `error-stack`.
2. or extend the `FrameKind` enumeration with `FrameKind::MultiContext` to hold not only one but multiple contexts.

In both cases, we could provide easy conversions from `Vec<Report<C>>` to build a multi-context.

Both proposals are breaking changes.

---

Thanks to @indietyp for [maintaining a list or expected changes](https://github.com/hashintel/hash/discussions/2377#discussioncomment-5566248)

**Comment by indietyp** [comment]
As a big proponent of related errors (after all, I was the one that first implemented them 😅), I'd like to give some context and my view/opinion on the matter.

I agree with @TimDiekmann that the potential for related errors is significant but may have untapped potential, but I am not in favor of any of the outlined proposals.

> It's not possible to convert a Report into Error as there may be multiple sources of an error, which the Error trait does not support.

I think this is by far the greatest issue, then again we have solved this through the `into_error` functions.

> `Report::current_context` only returns the first context, not a list of contexts, if multiple contexts at the same level exist. (The current implementation contains a bug in addition insofar that if adding a context and an attachment with the same type the attachment may be returned instead)

Yes, but I don't think that this is a problem of related errors, more like that the API is currently in between pre and post related errors.

> `Report::frames` and similar functions return a linearized iterator of all frames in the order in which they would be printed, however, there is no real distinction between a related error and a non-related error

This is still from the `0.1` to `0.2` switch, I am painfully aware that this isn't the correct behavior, maybe the visitor pattern would help here instead. (or one would only iterate until multiple contexts are encountered)

> Generally, every function needs to take a potentially related context into account

I think this is generally just a thing that happens with nested values, same applies to things like `XML` or `JSON`.

> Constructing "related-errors" is hard as we want to ensure, that at least one context is available. Thus, FromIterator cannot be implemented on Report.

This is something I am painfully aware of, but something that I at least locally have several fixes for. By implementing `fold_report()` on iterators and tuples it is possible to eliminate most of these problems. Currently, quite a lot of glue code is required that would be eliminated by them, an example for tuples can be seen in #1875.

Both solutions outlined are lacking because, IMO they either break encapsulation or hinder functionality (a lot). They only allow us to attach multiple contexts, but not create a real tree of reports, with independent attachments, but in a real-world application whenever there are independent computations, we'd like to aggregate we take multiple reports and smoosh them together. This means that we have two completely independent reports that are added together. This can happen multiple times, building up a report meaning that these multiple contexts might be deep down in the tree. Now I need to know/test if there have been multiple reports instead of a single API call.

These would practically reduce the functionality to the one present in miette, the error-handling library I "escaped" from, because of these exact reasons D:

In my opinion, it would make more sense to go more in the tree direction, instead of trying to fit a tree-sized piece in a stack-shaped hole. I think that the biggest problems for related errors in error-stack are:

1. tooling, how to construct them
2. visibility, make sure people know how to use them
3. consistency, make sure the API is consistent and not in `0.1`-`0.2` nirvana
4. usability, make sure surrounding calls like `sources()` are easier to use.

Do I have a specific solution in mind right now? Partially.

I think for (1) we can already do a lot more with methods exposed for iterators and tuples (`fold_report`). I think for (2) we need to explore how to maybe write the `lib.rs` to make it stand out more and/or create dedicated examples. (3) `current_contexts` should return `&[T]` instead of `&T`, ideally `NonEmptySlice<T>`, same with `sources()` and co. (4) for things like the iterator maybe it would make sense to switch up the pattern used (`Visitor` pattern) or create an iterator that results in multiple other iterators upon a split.

**Reply by TimDiekmann** [reply]
> I think this is by far the greatest issue, then again we have solved this through the into_error functions.

I don't think this is true. While we have an `into_error` function, this does not construct a full `Error`, but only a thin wrapper around a `Report` without implementing the `source` method. A lot of information is lost in this function.

> Yes, but I don't think that this is a problem of related errors, more like that the API is currently in between pre and post related errors.

How is this not related to Related-Errors? If a `Report>C>` contains a list of sources (constructed via the Releated-Errors API) and you call `Report::current_context`, only exactly one `C` is returned. The expected behavior, however, is to get a list of context.

> This is still from the 0.1 to 0.2 switch, I am painfully aware that this isn't the correct behavior, maybe the visitor pattern would help here instead. (or one would only iterate until multiple contexts are encountered)

I don't see how the visitor pattern would help here. It would still visit them in a linearized way. If the visitor distinguishes between a single frame and a set of frames, this is basically as returning `Frame::MultiContext` (second proposal)

> I think this is generally just a thing that happens with nested values, same applies to things like XML or JSON.

I hesitate to compare `error-stack` with `XML` or `JSON`. The most common use case is to call a function and then either attach something or change the context in the calling function while in `XML` or `JSON` it's very common to have deeply nested types with more than a single field. For the common `error-stack` use case, we still have nested fields but with a breadth of `1`. Only when adding more breadth to that, we encounter that issue.

I don't say my proposals are the only solution but it's very clear to me that we _need_ to change _something_ here. All issues would be resolved if it's possible to make the API more aware of a context batch. Generally, make the API consumer aware if they are dealing with a single `Context` or more than one.

> 1. tooling, how to construct them

Yes, the tooling currently is bad, but not the main issue with the current API. I agree that this is suboptimal and can be improved. There are several approaches we can try out (e.g. tuples or an extension to iterators/streams/sinks), but making this easier is _just_ an additional layer on the current API. This proposal is not based on the issue of how to construct this kind of `Report`s, but on the consumption of `Report`s.

> 2. visibility, make sure people know how to use them

Without proper tooling, I don't see why we should make this API more visible.

> 3, consistency, make sure the API is consistent and not in `0.1`-`0.2` nirvana

Surely `current_context` needs to be fixed, but this does not solve the issue around it

> 4. usability, make sure surrounding calls like `sources()` are easier to use.

My proposal is exactly about this point. To emphasize: **I don't want to remove the current functionality about Related Errors**, I'm a huge fan of it. I like to improve the API interface around it. Specifically, I want to make the type system aware of the possibility that a function may or may not return multiple errors. This should implicitly also improve the construction of those errors.

You mentioned `NonEmptySlice`, what is different from a `Vec<C>` or, as a proposed alternative, `MultiContext` (besides "slice" is a borrowed form)?

The main pain point currently is that with code like `report.attach(...).change_context(...)` I 100% know, that I only ever have a stack breadth of `1`, but it's not reflected in `Report`'s API at all. Unless I want to rely on `unsafe`, I have to add runtime checks as soon as I want to call `report.sources()` or (with the changes) `report.current_context()`. Especially, for the latter, this is really painful as this is an API that seems to be used quite often. I appreciate that this is probably fine for `sources`, but for `current_context` this is annoying because I know the type as I just changed the context. Currently, there is no way for a function to hint the caller, that it may return a potential set of errors, everything has to be done at runtime. If a function would return `Report<Vec<MyError>>` as the `Err` variant that is clear.

**Reply by indietyp** [reply]
I see where the disconnect is, I thought that what you wanted to do is something akin to:

![image](https://user-images.githubusercontent.com/7252775/230786597-c46481b4-ef0a-4814-bd06-38ec42c848a1.png)

which would destroy the "tree" aspect error-stack, I now realize that this likely isn't the case, but that you would like to instead express those constraints - if there - using the type system? That makes total sense, but more questions arise, what happens if I traverse the tree, now I need to know if the next context is either `1` or multiple, or would we have something like a `to_single()` function? Or would we just say that traversal is only possible in "multi" mode? (once we leave the `Report` and go into `Frame` territory)?

What about using const generics in that case? But that would mean that `extend_one` and `extend()` are no longer really possible.

The idea behind the `NonEmptySlice` would be something similar to [`non_empty_vec`](https://docs.rs/non-empty-vec/latest/non_empty_vec/struct.NonEmpty.html), where `first()` and such do not return `Option<C>`, but only `C`.

Please let me know if my ideas are somewhat incoherent and if I need to clarify something, so so many questions, and am excited about the prospect.

**Reply by TimDiekmann** [reply]
I would probably just stick with the implementation we currently have, so e.g. `sources: Box<[Frame]>` (actually I tried a different approach which did not work without sacrificing functionality). Thinking more about this it would probably make sense to extend `FrameKind` and have something like `Report<Vec<C>>`, so if you chose to use the `Frame` functions you can actually see a different but the high-level interface would be tied to the generic of `Report`. As long as we have the full control over `Report`, we can still assume, that `C: Context` is always given.

> What about using const generics in that case?

This won't work, no, as you cannot extend dynamically. Supporting fixed-size arrays, however, could be useful. One advantage in not using `Vec` but a more generic struct (I guess everything where `C: Iterator<Item: Context>`?) is, that we could easily use both and I doubt we need more functionality (besides constructing) where we need more than a simple iterator. 

> where `first()` and such do not return `Option<C>`, but only `C`.

Well, we would deviate from the normal `Iterator` interface here I guess. If this could be determined at compile time I'd highly favor that (i.e. decide on the return type based on `T` in `Report<T>`). We wouldn't be able to re-use signatures (as `sources`) but as they would be either singular or plural this is fine as the signatures would be `current_context() -> Option<C>` or `current_contexts() -> impl Iterator<Item = C>` and would never appear on the same `Report`.

I'd prefer anything which does not require additional helper types without sacrificing functionality. Also, it should be easy to convert `Report<C>` to `Report<Vec<C>>` (and a failable conversion in the other direction)

**Reply by indietyp** [reply]
To summarize into action items:

1) change nothing to how everything works (e.g. `Box<[Frame]>`)
2) split into a low-level interface (LL, `Frame`s) and high-level interface (HL, `Report`), LL stays the same, you will only be able to query for multiple sources, etc, HL overhaul to separate single `Context` vs. multiple `Context`

Separation between `Report<C>` and `Report<[C]>`

## `Report<C>` methods:

* `current_context() -> &C`
* `source() -> &Frame`
* `into_multiple() -> Report<[C]>`
* `new() -> Report<C>`

## `Report<[C]>` methods

* `into_single() -> Option<Report<C>>` (potentially `Result<Report<C>, Report<[C]>>` to be able to recover)
* `current_contexts() -> NonEmptySlice<C>`
* `sources() -> NonEmptySlice<Frame>`
* trait impl `Extend`
* `extend_one<T>(&mut self, Report<T>)`
* return value of of `Iterator::fold_results()`, `Tuple::fold_results()`
* `from_vec(reports: Report<C>) -> Report<[C]>`

## Shared methods

* `with(Report<C>) -> Report<[C]>`
* `contexts() -> NonZeroUsize` (the current amount of "splits")
* `change_context(context: C) -> Report<C>`
* `attach_*(self) -> Self`
---

I got a prototype working that does most of this pretty fast that is minimally invasive for maximum profit™

**Reply by indietyp** [reply]
By still facilitating internally a `Vec<Frame>` we're able to retain the ability of attaching attachments if we have multiple contexts. Internally nothing would change (we would just use the type-system to uphold all guarantees).

**Reply by indietyp** [reply]
We will still need to rework `frames()` and `frames_mut()` in their current form; they are useless either way once you start using related errors. I could imagine that we return an iterator, that once exhausted returns multiple other iterators.

**Reply by TimDiekmann** [reply]
Thanks, that generally looks good! I think we could use this possibility and clean-up the existing interface.

First of all, I like `Report<[C]>`. Probably not an issue, but this requires a `?Sized` bound.
I have a few comments on the provided list:
- `into_multiple`: I think having this is good, but I don't really like the name. What about this:
    - `impl<C: Context> From<Report<C>> for Report<[C]>`
    - `impl<C: Context> TryFrom<Report[C]> for Report<C>` with `Error = Report<[C]>`
- I just noticed that we don't have `source` on `Report`, only on `Frame`. If we add `source` to `Report` directly we, however, need to make it optional.
- We have `current_context` on singular reports, introducing `NonEmptySlice`, which would come with a whole new iterator interface, is greatly out of scope of `error-stack`. Generally, this will make interfacing `error-stack` with related errors a lot harder. Returning `ExactSizeIterator` (and `TrustedLen` on nightly) will be the better option here.
- instead of adding `Iterator::fold_results`, `Tuple::fold_results`, and `from_vec` (which actually is not a `Vec`) I'd provide `fn from_iter(impl IntoIterator<Item = Report<C>>) -> Option<Report<[C]>>`
- `with` seems to be unneeded (it's misleading and can easily be achieved by `into_multiple().extend_one()`, `Report` is not supposed to be a builder)
- `contexts` should be `len` or `num_contexts`, but `contexts` alone is misleading

**Reply by TimDiekmann** [reply]
This, however, does not yet address the first issue

**Reply by TimDiekmann** [reply]
Let's try to stick with the most minimal but correct interface and build more convenient functions on top of that.

**Reply by indietyp** [reply]
You are entirely correct, I was thinking of returning a slice, but the contexts won't share the same memory, so it's nonsense. An iterator would be the correct way to do things.

I think `Iterator::fold_results` and `Tuple::fold_results` are still valuable (I can go into more detail as to why if wanted), but something we can add after the overhaul, big 👍 on `fn from_iter()`.

Names are ofc temporary, am bad with names; maybe our local naming genius™ @Alfred-Mountfield, has a better idea.

We have `current_frames` instead of `source`; IMO, it would make sense to rename `current_frames` to `source`/`sources` respectively.

**Reply by TimDiekmann** [reply]
`current_frame` also returns the, well, current frame 😄, including attachments so it's different and this is basically the entry point to the internal `error-stack` API.

I agree that `fold_results` may be useful. I'm also thinking about abusing the `report!` macro here, currently, it's not really useful in any way.

> [I] am bad with names

I doubt that and I know @nonparibus agrees with me here. 🙃

**Comment by indietyp** [comment]
Moving this up so that our discussion can stay focused

---

1) change nothing to how everything works (e.g. `Box<[Frame]>`)
2) split into a low-level interface (LL, `Frame`s) and high-level interface (HL, `Report`), LL stays the same, you will only be able to query for multiple sources, etc, HL overhaul to separate single `Context` vs. multiple `Context`

Separation between `Report<C>` and `Report<[C]>`

## `Report<C>` and `Report<[C]>` methods:

* `attach_*(self) -> Self`
* `change_context(context: C) -> Report<C>`

## `Report<C>` methods:

* `new() -> Report<C>`
* `current_context() -> &C`
* `source() -> Option<&Frame>`
* `into_multiple() -> Report<[C]>`, name TBD
* `impl TryFrom<Report<[C]>>`

## `Report<[C]>` methods

* `into_single() -> Option<Report<C>>` (potentially `Result<Report<C>, Report<[C]>>` to be able to recover, name TBD)
* `current_contexts() -> impl ExactSizeIterator<C> + TrustedLen`
* `sources() ->&[Frame]`
* `from_iter(reports: impl Iterator<Item = Report<C>>) -> Option<Report<[C]>>`
* `num_contexts() -> NonZeroUsize` (the current amount of "splits"), name TBD
* `impl Extend<Report<C>>`
* `impl From<Report<C>>`

## Additional methods for constructing `Report<[C]>`
* `Iterator::fold_results()`, `Tuple::fold_results()` (after overhaul)
* Adjust macros

**Reply by indietyp** [reply]
I think that dedicated methods (especially when chaining) for conversion are quite valuable, otherwise one needlessly creates noise when converting between multiple and single

**Reply by TimDiekmann** [reply]
I have a POC working as well and I agree. Given the test we have:
```rust
#[test]
fn extend() {
    let mut err1: Report<[_]> = report!(Error).attach_printable("Not Supported").into();
    let err2 = report!(Error).attach_printable("Not Supported");
    let err3 = report!(Error).attach_printable("Not Supported");

    err1.extend([err2, err3]);
    assert_eq!(err1.current_frames().len(), 3);
    assert_eq!(err1.frames().count(), expect_count(2) * 3);
}
```

a dedicated function would be helpful. However, I wonder how much this is needed if we have `from_iter`

One could argue that the current implementation of `Fromiterator` is enough, it just looks a little bit weird:
```rust
let report = [report!(Error), report!(Error), report!(Error)]
    .into_iter()
    .collect::<Option<_>>();
```
Rust (and CLion) correctly resolves this to `Option<Report<[Error]>>`

**Reply by TimDiekmann** [reply]
Could you update this top-level comment as we proceed?

**Reply by TimDiekmann** [reply]
One observation I had with having `?Sized` bounds is that our implementation in fact has an implementation for negative trait bounds, which is _really_ helpful. 
```rust
impl<T> Report<T> {
    // Only functions for `Report<C>`, not `Report<[C]>`
}

impl<T> Report<[T]> {
    // Only functions for `Report<[C]>`, not `Report<C>`
}

impl<T: ?Sized> Report<T> {
    // Shared functions for `Report<C>` and `Report<[C]>`
}
```

**Reply by TimDiekmann** [reply]
I wonder if `new` should be shared as well, but probably this is not easily possible and instead `Report::new(context).into_multiple` should be used instead.

**Reply by indietyp** [reply]
I think only allowing `new()` for `Report<C>` also makes the distinction of multiple contexts vs single context clearer to the end user.

**Reply by TimDiekmann** [reply]
I rearranged the list a bit. Also, i moved `num_contexts` to `Report<[C]>`

**Reply by indietyp** [reply]
I think you misunderstood me, sorry! `fold_report` would complement `from_iter()` and would only be implemented on `Iterator<Item = Result<T, C: ?Sized>>` the result would be: `Result<impl FromIterator, [C]>`.

This would allow us to easily merge the results of an iterator into a single report.

**Reply by TimDiekmann** [reply]
I see, thanks for the clarification. On what you would implement `fold_results`? On `Report` itself?
Do we really want `C: ?Sized`? this implicitly would flatten the reports, which might be unexpected. Should we allow implicit conversion here generally/should we use `Iterator<Item = Result<T, impl Into<Report<C>>>>`?

Generally, we should never write `C: ?Sized` but use another generic name (e.g. `E` or `T`) as a context is never unsized.

**Reply by indietyp** [reply]
You are entirely correct, `C: ?Sized` is a dumb idea, `?Sized` should only be an implementation detail and would likely confuse the user.

I think we would solve this via an `IteratorExt` and `TupleExt` trait. That way, you're able to do: `iter.map(...).fold(...).fold_reports()`

This also means that in the case of `TupleExt`, we're able to retain all types, e.g.

```
// t1, t2, t3 are all results of fallible *independent* computations
let t1: Report<C> = f1();
let t2: Report<C> = f2();
let t3: Report<C>  = f3();

let (t1, t2, t3) = (t1, t2, t3).fold_reports().change_context(Error)?;
```

Which is very handy!

> Should we allow implicit conversion here generally/should we use `Iterator<Item = Result<T, impl Into<Report<C>>>>`?

That is a tricky question to answer; my gut feeling is to implement this for `Iterator<Item = Result<T, impl Into<Report<[C]>>>>`, which would allow implicit flattening. I think this is fine because for that to take place, the context needs to be the same anyway; this also allows for cases where in the `.map(...)` or other calls, we might have other computations that return `Report<[C]>` because they have some independent computations taking place and are already aggregating reports.

While thinking about it, should we also add `into_multiple`, and `into_single` (name TBD) to `ResultExt`?

**Reply by TimDiekmann** [reply]
> `Iterator<Item = Result<T, impl Into<Report<[C]>>>>`

I specifically asked about the singular variant to utilize the current `From` implementation. I'm unsure, if we should implement `From<C> for Report<[C]>` as well as with this change you cannot write `let r: Result<_, Report<_>> = x.into()` anymore (I guess?).

> While thinking about it, should we also add into_multiple, and into_single (name TBD) to ResultExt?

Probably we shouldn't (yet). The traits are polluting the global `Result` and `Future` namespace already. To keep the option to add it later we however should consider sealing it.

**Reply by indietyp** [reply]
I was thinking about `From<Report<C>> for Report<[C]>`, so your iterator can both use `Report<C>` and `Report<[C]>`

**Comment by TimDiekmann** [comment]
Done in https://github.com/hashintel/hash/pull/5047
